Make llb.ReadonlyRootFS usable with common container images

ijc opened the pull request:
busybox image):

    OCI runtime create failed: container_linux.go:348:
    starting container process caused "process_linux.go:402:
    container init caused \"rootfs_linux.go:58:
    mounting \\\"proc\\\" to rootfs \\\"/.../rootfs\\\" at \\\"/proc\\\"
    caused \\\"mkdir /.../rootfs/proc: read-only file system\\\"\"": unknown

This is because we were setting the underlying snapshot readonly so the various
mountpoints (here /proc) cannot be created. This would not be necessary if
those mountpoints were present in images but they typically are not.

The right way to get around this (used e.g. by `ctr`) is to use a writeable
snapshot but to set root readonly in the OCI spec. In this configuration the
rootfs is writeable when mounts are processed but is then made readonly by the
runtime (runc) just before entering the user specified binary within the
container.

This involved a surprising amount of plumbing.

Use this new found ability in the dockerfile converter's `dispatchCopy`
function.

Signed-off-by: Ian Campbell <ijc@docker.com>

I uses #323 to regenerate the protobuf but didn't include it in this PR.

ijc commented:
CI fail was due to `worker/runc/runc_test.go` which I hadn't run due to 
```
// +build linux,no_containerd_worker
```
not sure why the test is tagged, could it not be part of the usual suit?

Anyway, will fix.

tonistiigi commented:
> not sure why the test is tagged, could it not be part of the usual suit?

@AkihiroSuda This should be reversed, right? So builds by default but can be optionally disabled with a build tag.

AkihiroSuda commented:
>  This should be reversed, right?

Yes, IIUC

ijc commented:
> Why do we need a new field for this instead of detecting if a mount with readonly set is pb.RootMount in llbop/exec.

It seemed less flexible is all, in principal someone might want to use `pb.RootMount` with a real readonly snapshot rather than an oci spec level readonlyness. containerd has the same split I think for the same reason. It could be argued that this flexibility is not useful in the context of buildkit I suppose.

> For the executor, the readonly bool can be part of executor.Meta.

I had it plumbed through in various layers of `Meta` but it looked odd in the higher level ones (it doesn't really fit with the other fields). Will change it to be just at the executor level

ijc commented:
I added a commit (not intended for eventual squashing) which inverted the tag in `runc_test.go` from `no_containerd_worker` to `!no_runc_worker`.

I also added a commit (intended for eventual squashing) which moved the readonly parameter to the various `Executor`-like interfaces to `executor.Meta`. I didn't go further up the call chain for now.

tonistiigi commented:
@AkihiroSuda wdyt? I think for the builder there probably isn't a valid case for having a duplicate way to set read-only to exec and to the root mount itself. It also leaves broken cases like when you set the mount read-only but don't set readonlyrootfs. If we allow to actually use `ImmutableRef` as root there is a chance that a future runc change will start to make some builds fail as there is no guarantee that runc can accept a read-only rootfs.

ijc commented:
I've been playing with removing the new field and instead using the `m.Readonly` for the single `m` where `m.Dest == pb.RootMount` instead.

There are a bunch of existing places with logic which special cases `m.Readonly` vs `m.Dest == pb.RootMount` and I've found at least one more place to adjust but I still can't get it working. The failure case is related to my change to `dispatchCopy` in this PR and the symptom is that when `llb.ReadonlyRootFS()` is used there the output contains the rootfs of the copy container and not `dest`.  The upshot is that `TestIntegration/TestDockerfileAddArchive` fails (for both workers).

I suspect I must be introducing an off-by-one in the `inputs` or `outputs` array handling in `execOp.Run` or somewhere but I can't for the life of me spot it, or any other problem.

I've pushed the current state to https://github.com/moby/buildkit/compare/5b499d557b7ee574ba906ffd9e63b419b2c59a84...ijc:readonly-rootfs-execop-wip &mdash; any extra eyes or tips would be appreciated!

ijc commented:
I knew that I'd spot it as soon as I hit "Comment"! The issue is in `getMountIndexFn`...

tonistiigi commented on client/llb/exec.go:
```diff
@@ -24,11 +24,7 @@ func NewExecOp(root Output, meta Meta, readOnly bool, md OpMetadata) *ExecOp {
-	} else {
-		e.root = &output{vertex: e, getIndex: e.getMountIndexFn(rootMount)}
-	}
+	e.root = &output{vertex: e, getIndex: e.getMountIndexFn(rootMount)}
```
Is this correct though? Shouldn't we handle it in `llbop/exec` instead? Skipping this output would make a sequence of execs with readonly rootfs to run in parallel. So in `llbop/exec` don't add it to the outputs and just release the mutableref as soon as the exec has completed.

ijc replied:
I was considering highlighting this particular portion for extra maintainer scrutiny since I@m really not sure ;-).

My thinking was that when readOnly was true this was actually being subverted further down so we wanted to treat it as writeable here, it ties in with the change just below at https://github.com/moby/buildkit/pull/324/files#diff-b6e1be4db0a6b8d1aada0afaaa76287dR155 which adds `m.target == pb.RootMount` to the output index stuff.

I'll take a look at the `llbop/exec` portions you suggest and see if I can make it work since that would keep the internals internal.

AkihiroSuda commented:
I think we should try to keep `pb.ExecOp` fields immutable

tonistiigi commented:
@AkihiroSuda wdym? I'm not suggesting mutating the `pb.ExecOp` in `llbop/exec`, only filtering the outputs `Exec` returns. Looking more at the code, I guess I just don't understand why the `getMountIndexFn` change (and https://github.com/moby/buildkit/pull/324/files#diff-b6e1be4db0a6b8d1aada0afaaa76287dR155) was needed. It doesn't look correct to me though because readonly mount shouldn't create an output, even if it is root. Instead, I think the issue is that `llbop/exec` only looks if mount has an output while deciding to create a mutable ref - it should create a mutable ref for a root mount even if it doesn't have an output.

AkihiroSuda commented:
My comment above was for https://github.com/moby/buildkit/pull/324/files#diff-d086cb6f36687838804d1974ad60ea21R114 😅

ijc commented:
I guess @tonistiigi's suggestion to do this all in `llbop/exec` might make that mutation go away, I'll take a look and see.

ijc commented:
That force push was a straight forward rebase onto current `master`, I'll do the `llbop/exec` thing next.

ijc commented:
Moved everything into `llbop/exec`, the result is indeed much nicer, thanks!

ijc commented on frontend/gateway/gateway.go:
```diff
@@ -64,6 +64,7 @@ func (gf *gatewayFrontend) Solve(ctx context.Context, llbBridge frontend.Fronten
 	_, isDevel := opts[keyDevel]
 	var img ocispec.Image
 	var rootFS cache.ImmutableRef
+	readonly := true // XXX could this be set to false for all cases below?
```
I cannot at all remember what I meant by this comment, since setting the containers to readonly is surely more desirable when possible.

Using readonly doesn't appear to have caused any problems with the test suite, but I don't know how good the test coverage is here (esp. for `gateway-devel` mode). @tonistiigi shall I drop the comment or switch to writable rootfs do you think?

ijc commented on solver/llbop/exec.go:
```diff
@@ -102,6 +103,18 @@ func (e *execOp) Run(ctx context.Context, inputs []solver.Ref) ([]solver.Ref, er
+			readonlyRootFS = m.Readonly
+			if m.Output == pb.SkipOutput && readonlyRootFS {
+				// XXX this duplicates a case from above.
+				active, err := e.cm.New(ctx, ref, cache.WithDescription(fmt.Sprintf("mount %s from exec %s", m.Dest, strings.Join(e.op.Meta.Args, " ")))) // TODO: should be method
```
I feel a bit bad duplicating this code here. The comment says `TODO: should be method` but I'm not too sure what the method was intended to be on. Perhaps on `cache.ImmutableRef` (i.e. `ref` here)?

Or maybe just a freestanding helper func (or even a local more `foo := func(...)` closure thing) would be better?

ijc commented:
I had to rebase over a conflict in `client_test.go` so I took the opportunity to squash. I also:
* made the gateway containers writeable again
* added a commit to remove the code duplication discussed above.
